cmdstanpy/stanfit.py

```diff
@@ -1397,24 +1401,56 @@ def stan_variable(
             raise ValueError('no variable name specified.')
         if var not in self._metadata.stan_vars_dims:
             raise ValueError('unknown variable name: {}'.format(var))
+        if warn and inc_iterations and not self._save_iterations:
+            get_logger().warning(
+                'Intermediate iterations not saved to CSV output file. '
+                'Rerun the optimize method with "save_iterations=True".'
+            )
         if warn and not self.runset._check_retcodes():
             get_logger().warning(
                 'Invalid estimate, optimization failed to converge.'
             )
 
-        col_idxs = list(self._metadata.stan_vars_cols[var])
-        vals = list(self._mle)
-        xs = [vals[x] for x in col_idxs]
-        shape: Tuple[int, ...] = ()
+        col_idxs = self._metadata.stan_vars_cols[var]
+        if inc_iterations and self._save_iterations:
+            num_rows = self._all_iters.shape[0]
+        else:
+            num_rows = 1
+
+        # extract and reshape, container var
         if len(col_idxs) > 0:
-            shape = self._metadata.stan_vars_dims[var]
-        return np.array(xs).reshape(shape)
+            dims = (num_rows,) + self._metadata.stan_vars_dims[var]
+            # pylint: disable=redundant-keyword-arg
+            if num_rows > 1:
+                return self._all_iters[:, col_idxs].reshape(  # type: ignore
+                    dims, order='F'
+                )
+            else:
+                mle = np.expand_dims(self._mle, axis=0)  # hack for col indexing
+                return (
+                    mle[0, col_idxs]
+                    .reshape(dims, order='F')  # type: ignore
+                    .squeeze(axis=0)
+                )
```

Comment:
What is the original shape for mle? and what is the output shape?

Reply:
in Stan, a multi-dim variable is output by the Stan 'write_array' function column major, i.e., Fortran order - cf the header row from test model `multidim_vars.stan` -

```
rep.1.1.1,y_rep.2.1.1,y_rep.3.1.1,y_rep.4.1.1,y_rep.5.1.1,y_rep.1.2.1,y_rep.2.2.1,y_rep.3.2.1,y_rep.4.2.1,y_rep.5.2.1,y_rep.1.3.1,y_rep.2.3.1,y_rep.3.3.1,y_rep.4.3.1,y_rep.5.3.1,y_rep.1.4.1,y_rep.2.4.1,y_rep.3.4.1,y_rep.4.4.1,y_rep.5.4.1,y_rep.1.1.2,y_rep.2.1.2,y_rep.3.1.2,y_rep.4.1.2,y_rep.5.1.2,y_rep.1.2.2,y_rep.2.2.2,y_rep.3.2.2,y_rep.4.2.2,y_rep.5.2.2,y_rep.1.3.2,y_rep.2.3.2,y_rep.3.3.2,y_rep.4.3.2,y_rep.5.3.2,y_rep.1.4.2,y_rep.2.4.2,y_rep.3.4.2,y_rep.4.4.2,y_rep.5.4.2,y_rep.1.1.3,y_rep.2.1.3,y_rep.3.1.3,y_rep.4.1.3,y_rep.5.1.3,y_rep.1.2.3,y_rep.2.2.3,y_rep.3.2.3,y_rep.4.2.3,y_rep.5.2.3,y_rep.1.3.3,y_rep.2.3.3,y_rep.3.3.3,y_rep.4.3.3,y_rep.5.3.3,y_rep.1.4.3,y_rep.2.4.3,y_rep.3.4.3,y_rep.4.4.3,y_rep.5.4.3,
```

I'll check to see if we need to add more tests to make sure this is really working as expected